[Misc] IO Processor plugins for pooling models

christian-pinto opened the pull request:
This PR introduces the ability of pre/post processing pooling models input/output via an IO Processor plugin. This allows users to feed complex data structures to vLLM that are then parsed into prompts by the plugin, and generate any type of data out of the model output (e.g., images).

Main features:

- A new endpoint for offline inference `encode_with_io_processor` added to `LLM` and `AsyncLLM` for offline inference, and a  new endpoint `/io_processor_pooling` for online serving.
- A new plugin group `vllm.io_processor_plugins`.
- IO Processor Plugins are instantiated on a per instance basis via an `EngineArgs` argument `io_processor_plugin` or via a custom field (`io_processor_plugin`) in the model configuration (config.json).
- Example plugin for the Prithvi model included in the testing pipeline as well as examples for both offline inference and online serving.
- Documentation updated

This is an alternative implementation to #21621.

@DarkLight1337 @maxdebayser @mgazz 

CLOSE https://github.com/vllm-project/vllm/pull/12249

christian-pinto committed:
multimodal processor plugins for image tiling

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Some minor fixes

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed pre-commit hooks errors

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/entrypoints/llm_for_image_tiling.py:
```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+from vllm import LLM
```
I would move this to the example script as well, since this can't be used in online inference. For offline inference, users can simply copy from the example script.

DarkLight1337 replied:
To avoid adding bloat to our library, I much prefer these types of wrappers to be the user's responsibility

DarkLight1337 commented on vllm/plugins/multimodal_data_processors/__init__.py:
```diff
@@ -0,0 +1,59 @@
+logger = logging.getLogger(__name__)
+
+
+def get_multimodal_data_processor(
```
Based on the discussion from https://github.com/vllm-project/vllm/pull/14526, I think we can apply the processor to `encode` method specifically, while leaving the other methods unchanged. This would allow it to work in online inference as well.

christian-pinto committed:
Added new entrypoint for online inference

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -439,6 +442,10 @@ def translation(request: Request) -> OpenAIServingTranslation:
     return request.app.state.openai_serving_translation
 
 
+def images_prediction(request: Request) -> ServingImagesPrediction:
```
TBH I have structured it like this after discussing with you in the other PR (https://github.com/vllm-project/vllm/pull/21621#issuecomment-3131778023). You were suggesting avoiding touching the async engine and instead put everything in the entrypoint.

I started liking the entrypoint idea because as models show up we can then start offering `images` endpoints instead of relying on the basic pooling one.

Also, if I do this in the encode, I will have to initialize the plugin in the AsyncLLM init  and therefore add an extra argument for starting an engine with/without such data pre/post processor loaded. 

Just making sure we are fully aligned before I try implementing it inside the encode.

DarkLight1337 replied:
Oh, I was mainly concerned about the aggregation part, where multiple inputs correspond to one output. After looking at #14526, I feel that we still need a plugin system to postprocess the outputs (in a one-to-one manner). Sorry for the confusion!

christian-pinto replied:
I see. In my case the plugin is what splits the input into multiple prompts

pre-process the input with the plugin -> generate 1+ prompts -> inference -> postprocess 1+ outputs with plugin -> 1 final output

Could we make one plugin system to serve both purposes? In the end a plugin could either aggregate multiple inference results as well as simply do a 1-to-1. Especially is we overhaul the PoolerOutput data structure we should be able to cater for multiple use-cases.

DarkLight1337 replied:
Since the existing methods assume a 1-to-1 correspondance between input and output, I think it would be better to define a new method (both in LLM class and online inference) for this flexibility. Maybe call it `LLM.plugin_forward`

christian-pinto committed:
Moved applying plugins to source /workspace/venvs/vllm/bin/activate

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Addressed CI failures

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed bug in offlne example

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Added simple test

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on tests/plugins_tests/test_multimodal_processor_plugins.py:
```diff
@@ -0,0 +1,15 @@
+
+
+def test_loading_missing_plugin(monkeypatch):
+    monkeypatch.setenv("VLLM_USE_MULTIMODAL_DATA_PROCESSOR_PLUGIN", "plugin")
```
Let's not limit this to multi-modal data. I would call this `VLLM_USE_OUTPUT_POSTPROCESSOR_PLUGIN`

DarkLight1337 replied:
Use "postprocessor" instead of "processor" to avoid confusing with `OutputProcessor` class

christian-pinto replied:
yes, makes sense. Would something like `IOProcessorPlugin` be more appropriate? in the end we are also offering the possibility of pre-processing the prompts.

DarkLight1337 replied:
Yes, that would also work

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -1145,6 +1151,71 @@ def encode(
+        lora_request: Optional[Union[list[LoRARequest], LoRARequest]] = None,
+        pooling_task: PoolingTask = "encode",
+        tokenization_kwargs: Optional[dict[str, Any]] = None,
+    ) -> list[PoolingRequestOutput]:
```
I think we can allow the output type to be arbitrarily defined by the plugin

DarkLight1337 commented on vllm/v1/engine/async_llm.py:
```diff
@@ -546,6 +551,71 @@ async def encode(
+                model_out=final_results,
+                request_id=request_id,
+            ))
+
+        if num_prompts > 1:
+            out_tensors = [
+                output.outputs.data for output in final_results if output
+            ]
+            stacked_tensor = torch.stack(out_tensors, dim=0)
+            request_out = PoolingRequestOutput(
+                outputs=PoolingOutput(data=stacked_tensor),
+                prompt_token_ids=[],
+                finished=True,
+                request_id=request_id)
+        else:
+            request_out = final_results[0]
+
+        request_out.task_output = post_processed_outputs[0]
+        return request_out
```
Same here, the plugin can define the outputs entirely

DarkLight1337 replied:
FastAPI should be able to convert the returned response automatically as long as the response is a pydantic model. Maybe the plugin can also define how to convert the output into a response.

christian-pinto replied:
You are right, I have done a quick test and endpoints can return the base `OpenAIBaseModel` and any of its implementations. However, if we go down the road of a single endpoint for any type of plugin, we will have to open the input side as well.

Basically, the input request will have to accept be any dict and the response will have to be anything of type `OpenAIBaseModel`.

Would there be any security implications in doing this? It feel like I can throw anything at vLLM and get anything out of it. It is fine in the offline fashion, where users control everything. However, for the online case I believe we should still define dedicated endpoints.

Say one endpoint for images generation that uses the below request and response types

```python
class ImagesPredictionRequest(OpenAIBaseModel):
    image: ImageData
    image_format: Literal["jpeg", "png", "tiff"]
    model: Optional[str]
    response_format: Literal["b64_json", "url"] = "b64_json"

class ImageData(OpenAIBaseModel):
    data: Optional[str]
    data_format: Literal["b64_json", "url"] = "b64_json"

class ImagesGenerationResponse(OpenAIBaseModel):
    created: int
    image: ImageData
    image_format: Optional[Literal["jpeg", "png", "tiff"]] = None

```

We can still have the plugin export a function that generates the response for online serving and at the endpoint we validate the format. With time we can add more endpoints if needed.

How's that sound?

DarkLight1337 replied:
Since the plugin is defined server-side, I think it's the responsibility of the plugin to correctly parse the input. So maybe we can have a parse_request function in the plugin.

DarkLight1337 reviewed (commented):
Sorry for the delay, was caught up with other things. Left some comments.

christian-pinto committed:
New version

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/entrypoints/openai/serving_pooling_with_io_plugin.py:
```diff
@@ -0,0 +1,71 @@
+logger = init_logger(__name__)
+
+
+class ServingPoolingWithIOPlugin(OpenAIServing):
```
Let's move this plugin to tests like the other plugins

DarkLight1337 replied:
Oh never mind, I see what you mean now. This is not the actual plugin class, it's the endpoint that serves the plugin.

DarkLight1337 commented on examples/online_serving/openai_translation_client.py:
```diff
@@ -68,7 +67,7 @@ def main():
     sync_openai(foscolo, client)
     # Run the asynchronous function
-    asyncio.run(stream_openai_response(foscolo, openai_api_base, openai_api_key))
+    # asyncio.run(stream_openai_response(foscolo, openai_api_base, openai_api_key))
```
Is this change intentional?

christian-pinto replied:
Definitely not. I will revert right away.

DarkLight1337 reviewed (commented):
Sorry I was working on other stuff at the same time so didn't read that thoroughly. Left another round of comments

christian-pinto committed:
Latest changes and added doc

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Typo in doc

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/plugins/io_processors/interface.py:
```diff
@@ -0,0 +1,60 @@
+        request_id: Optional[str] = None,
+        **kwargs,
+    ) -> Union[PromptType, Sequence[PromptType]]:
+        ...
```
We can default to non-async version

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -201,6 +202,7 @@ def __init__(
                                            CompilationConfig]] = None,
         logits_processors: Optional[list[Union[str,
                                                type[LogitsProcessor]]]] = None,
+        io_processor_plugin: Optional[str] = None,
```
Same, can also be extracted from `vllm_config`

christian-pinto replied:
all done .

christian-pinto committed:
Moved io-processor-plugin to vllm_config

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on vllm/plugins/io_processors/interface.py:
```diff
@@ -0,0 +1,58 @@
+        request_id: Optional[str] = None,
+        **kwargs,
+    ) -> Any:
+        self.post_process(model_out, request_id, **kwargs)
```
```suggestion
        return self.post_process(model_out, request_id, **kwargs)
```

DarkLight1337 commented on vllm/entrypoints/openai/serving_pooling_with_io_plugin.py:
```diff
@@ -0,0 +1,71 @@
+        except asyncio.CancelledError:
+            return self.create_error_response("Client disconnected")
+
+        io_processor = await self.engine_client.get_io_processor()
+        response = io_processor.plugin_out_to_response(output)
```
I think this should be handled by the engine client as well?

christian-pinto replied:
I see what you mean. My thinking here is that if someone is using the AsyncLLM from let's say ray or similar they would not need to convert the output to the `IOProcessorPluginResponse` type that we use only for the serving mode so no need to tun `plugin_out_to_response` in those cases. Does that make sense?

christian-pinto committed:
few more changes

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented:
@DarkLight1337 All requested changes applied PTAL

christian-pinto committed:
Fixed typos in doc

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed more typos in doc

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented on docs/design/io_processor_plugins.md:
```diff
@@ -0,0 +1,75 @@
+
+## Writing a IOProcessor Plugin
+
+IOProcessor plugins implement the interface of the same name (<gh-file:vllm/plugins/io_processors/interface.py>):
```
Fixed typos around, there were more that I hadn't noticed before.

DarkLight1337 commented on docs/design/io_processor_plugins.md:
```diff
@@ -0,0 +1,75 @@
+        raise NotImplementedError
+
+    @abstractmethod
+    def plugin_out_to_response(self,
+                               plugin_out: Any) -> IOProcessorPluginResponse:
```
```suggestion
    def output_to_response(self,
                           plugin_output: Any) -> IOProcessorPluginResponse:
```

I think `plugin` is a bit redundant given this class is already part of the plugin

DarkLight1337 replied:
Also maybe we can use type variables to indicate that the output of `post_process` is the input to `output_to_response`

christian-pinto replied:
I have removed the word plugin around where not necessary. As a rule I have kept it where it is linked with loading the plugin. When using it is just a IOProcessor.

christian-pinto committed:
Streamlined variables naming

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Some more renaming

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Updated IOProcessor interface in documentation

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto commented on docs/design/io_processor_plugins.md:
```diff
@@ -0,0 +1,75 @@
+
+    @abstractmethod
+    def plugin_out_to_response(self,
+                               plugin_out: Any) -> IOProcessorResponse:
```
the documentation, of course. Done!

DarkLight1337 commented on docs/design/io_processor_plugins.md:
```diff
@@ -0,0 +1,78 @@
+The `post_process*` methods take `PoolingRequestOutput` objects as input and generate a custom plugin output.
+An implementation of the `encode_with_io_processor` method is available [here](../../vllm/entrypoints/llm.py) and [here](../../vllm/v1/engine/async_llm.py).
+
+The `plugin_out_to_response` method is used only for online serving and converts the plugin output to the `IOProcessorResponse` type that is then returned by the API Server. The implementation of the `/io_processor_pooling` serving endpoint is [here](../../vllm/entrypoints/openai/serving_pooling_with_io_plugin.py).
```
```suggestion
The `output_to_response` method is used only for online serving and converts the plugin output to the `IOProcessorResponse` type that is then returned by the API Server. The implementation of the `/io_processor_pooling` serving endpoint is [here](../../vllm/entrypoints/openai/serving_pooling_with_io_plugin.py).
```

christian-pinto committed:
Some more changes in the docs

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Docs changes

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 reviewed (approved):
Thanks for your patience, I think the implementation looks good now.

But it would be nice to test the functionality of the plugin. Could you put your `prithvi_io_processor_plugin` plugin into `tests/plugins` and add some tests to validate it?

christian-pinto committed:
Doc

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented:
cc @maxdebayser please take another pass on this

christian-pinto committed:
Added prithvi io processor plugin test

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed plugin after pre-hooks pass

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

DarkLight1337 commented on tests/plugins_tests/test_io_processor_plugins.py:
```diff
@@ -0,0 +1,133 @@
+        hasattr(output, attr)
+        for attr in ["type", "format", "data", "request_id"])
+
+    # verify the output image in base64 is of the corerct length
```
```suggestion
    # verify the output image in base64 is of the correct length
```

DarkLight1337 commented on tests/plugins_tests/test_io_processor_plugins.py:
```diff
@@ -0,0 +1,133 @@
+        plugin_data.get(attr)
+        for attr in ["type", "format", "data", "request_id"])
+
+    # verify the output image in base64 is of the corerct length
```
```suggestion
    # verify the output image in base64 is of the correct length
```

christian-pinto committed:
Fixed typos

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

christian-pinto committed:
Fixed plugin correctness check

Signed-off-by: Christian Pinto <christian.pinto@ibm.com>

maxdebayser commented:
In general this is great.

One thing though is a bit odd: in the LLM -> Engine path the io processor is called in the LLM entrypoint. In the ApiServer -> AsyncEngine path the processor is called in the Async Engine. I think it should be at the Engine level in both cases.

Also, I would prefer if we added a `use_io_processor` flag to `encode` instead of creating a new method. Since `LLM.encode()` is the backend for several other entrypoint functions, it would allow us to add preprocessing plugins to other use cases.

Finally, I personally would prefer a simpler code for the plugin tests. Like adding some tokens to the request or multiplying the output. Something that takes just a couple of lines of code and can be tested easily. But this is just a matter of personal preference.

@christian-pinto , I know you're going to be OOO. If you agree with these suggestions, I'm happy to work on them in the same PR or a follow-up PR.

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Refactor the io processor integration

- Unify the llm and openai endpoints
- Move the processing logic to the entrypoint layer in
  the api server case for consistency
- Fix merge problems due to the deprecation of V0 Pooling
- Allow the IO processor the handle a generator in the async
  cases instead of a collected list to enable pipelining

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
@DarkLight1337 and @christian-pinto , I've implemented most of the changes I suggested above except for the simplification of the plugin. On the entrypoint -> engine consistency suggestion I actually changed my mind and moved the processing to the entrypoint in the APIServer case because the processor can create split the input into several requests and then aggregate them into a single response. There are several precedents for that at the entrypoint layer, but not at the engine client layer.

maxdebayser committed:
try to appease linter

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
appease linter

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
fix default code

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 reviewed (commented):
I think this is more streamlined now, thanks. Though you should update the documentation and examples as well

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
fix docs and examples

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented on vllm/engine/protocol.py:
```diff
@@ -229,6 +230,19 @@ def encode(
         """Generate outputs for a request from a pooling model."""
         ...
 
+    async def encode_with_io_processor(
```
Remove this?

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
remove leftovers

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -928,6 +935,22 @@ def encode(
+            validated_prompt = self.io_processor.parse_request(prompts)
+
+            # obtain the actual model prompts from the pre-processor
+            prompts = (self.io_processor.pre_process(prompt=validated_prompt))
```
```suggestion
            prompts = self.io_processor.pre_process(prompt=validated_prompt)
```
Nit

DarkLight1337 commented on tests/plugins/prithvi_io_processor_plugin/prithvi_io_processor/prithvi_processor.py:
```diff
@@ -0,0 +1,446 @@
+                        julian_day, "%m%d").timetuple().tm_yday)
+                temporal_coords.append([year, julian_day])
+        except Exception as e:
+            print(f"Could not extract timestamp for {file} ({e})")
```
Use logger?

maxdebayser committed:
address review comments

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

DarkLight1337 commented on tests/plugins/prithvi_io_processor_plugin/prithvi_io_processor/prithvi_processor.py:
```diff
@@ -220,7 +223,8 @@ def load_image(
                 temporal_coords.append([year, julian_day])
         except Exception as e:
-            print(f"Could not extract timestamp for {file} ({e})")
+            logger.exception("Could not extract timestamp for %s (%s)", file,
+                             str(e))
```
```suggestion
            logger.exception("Could not extract timestamp for %s", file)
```

The exception stack trace is logged automatically under `logger.exception` so no need to include it in the string again

maxdebayser replied:
I suspected this, but I found both on the code base. I'm removing it now.

maxdebayser committed:
address review comments

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
use vllm_runner to clean up memory

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

maxdebayser commented:
The failing blackwell test is not related.

maxdebayser committed:
Merge branch 'upstream_main' into multimodal_processing_plugin_entrypoint

Signed-off-by: Max de Bayser <mbayser@br.ibm.com>

christian-pinto commented:
Many thanks @maxdebayser for your contributions. The whole code is indeed more streamlined now. 
Thanks @DarkLight1337  for your patience in dealing with this PR.
I have spotted a few issues with the links in the docs. I will open a new PR to fix them.

On to the next one now!
